src/documentation/developer.rs

```diff
@@ -0,0 +1,803 @@
+  content: String,
+  /// The location of the start of this token in the source string, in bytes
+  location: usize
+}
+
+/// A string token from a source string with the location at which it occurs in the source string
+/// as line on which it occurs (1 indexed) and the column of its first character (0 indexed)
+#[derive(Debug)]
+struct TokenWithLineColumn {
+  /// The full contents of this token, including pre/post characters (like '//')
+  content: String,
+  /// The first line on which the token appears in the source file (1 indexed)
+  line: usize,
+  /// The column where the first character of this token appears in the source file (0 indexed)
+  column: usize
+}
+
+/// Is a token of type (developer) block comment, (developer) line comment or something else
+#[derive(Debug, Eq, PartialEq)]
+enum TokenType {
+  BlockComment,
+  LineComment,
+  Other
+}
+
+impl Display for TokenType {
+  fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+    let kind = match self {
+      TokenType::BlockComment => "developer block comment",
+      TokenType::LineComment => "developer line comment",
+      TokenType::Other => "not a developer comment"
+    };
+    write!(f, "{}", kind)
+  }
+}
+
+impl TokenType {
+  /// The prefix string for this type of token
+  fn pre(&self) -> &str {
+    match self {
+      TokenType::BlockComment => BLOCK_COMMENT_PREFIX,
+      TokenType::LineComment => LINE_COMMENT_PREFIX,
+      TokenType::Other => OTHER_PREFIX
+    }
+  }
+  /// The postfix string for this type of token
+  fn post(&self) -> &str {
+    match self {
+      TokenType::BlockComment => BLOCK_COMMENT_POSTFIX,
+      TokenType::LineComment => LINE_COMMENT_POSTFIX,
+      TokenType::Other => OTHER_POSTFIX
+    }
+  }
+  /// The length of the prefix for the token in characters
+  fn pre_in_chars(&self) -> usize {
+    self.pre().chars().count()
+  }
+  /// The length of the postfix for the token in characters
+  fn post_in_chars(&self) -> usize {
+    self.post().chars().count()
+  }
+}
+
+/// A token from a source string with its variant (TokenType) and the line and column on which it
+/// occurs according to the description for `TokenWithLineColumn`
+#[derive(Debug)]
+struct TokenWithType {
+  /// Is the token a block developer comment, line developer comment or something else
+  kind: TokenType,
+  /// The full contents of this token, including pre/post characters (like '//')
+  pub content: String,
+  /// The first line on which the token appears in the source file (1 indexed)  pub line: usize,
+  pub line: usize,
+  /// The column where the first character of this token appears in the source file (0 indexed)
+  pub column: usize,
+}
+
+impl TokenWithType {
+  /// Convert a `TokenWithLineColumn` to a `TokenWithType`. The kind is worked out from the content
+  /// by checking against the developer block comment & line comment regexes.
+  fn from (token: TokenWithLineColumn) -> Self {
+    let kind = {
```

Comment:
nit: The outermost `{ .. };` are not necessary.

Reply:
Removed in latest push. Matter of taste, personally I find it a bit easier to ready with the extra brackets.